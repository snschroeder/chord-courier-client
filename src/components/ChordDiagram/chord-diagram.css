.diagram {
  position: relative;
  height: 250px;
  width: 250px;
  margin: 0 auto;
  background-image: url('../../image-assets/diagram.png');
  background-size: contain;
  background-repeat: no-repeat;
  border-radius: 3px;
  margin-bottom: 12%;
  margin-top: 12%;
}

.open-diagram {
  position: relative;
  height: 254px;
  width: 250px;
  margin: 0 auto;
  background-image: url('../../image-assets/open-diagram.png');
  background-size: contain;
  background-repeat: no-repeat;
  border-radius: 3px;
  margin-bottom: 12%;
  margin-top: 12%;
}

.barre {
  background-image: url('../../image-assets/barre.png');
  background-size: contain;
  background-repeat: no-repeat;
  width: 250px;
  height: 30px;
  top: -52px;
  position: absolute;
}