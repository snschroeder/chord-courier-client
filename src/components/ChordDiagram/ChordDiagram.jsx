/* eslint-disable react/no-array-index-key */
/* eslint-disable array-callback-return */
/* eslint-disable no-undef */
import React from 'react';
import PropTypes from 'prop-types';
import MutedString from './muted-string/MutedString';
import OpenString from './open-string/OpenString';
import Finger from './finger/Finger';
import Fret from './fret/Fret';
import './chord-diagram.css';

export default function ChordDiagram(props) {
  const generateFingerPosition = (position) => {
    let fingerPos = position.split(',');
    const start = fingerPos[0];
    const startFret = start !== '0' ? 'diagram' : 'open-diagram';
    let barre = '';
    fingerPos = fingerPos.map((pos) => (!'xbo'.includes(pos) ? parseInt(pos, 10) : pos));

    // eslint-disable-next-line consistent-return
    fingerPos = fingerPos.map((pos, index) => {
      if (pos === 'b') {
        barre = 'barre';
      } if (pos === 'x') {
        return <MutedString stringNo={index} key={index} />;
      } if (pos === 'o') {
        return <OpenString stringNo={index} key={index} />;
      } if (index === 0 && pos !== 0) {
        return <Fret fret={pos} key={index} />;
      } if (index >= 1 && typeof pos === 'number') {
        return <Finger stringNo={index} fret={pos} key={index} />;
      }
    });
    return { startFret, fingerPos, barre };
  };

  const { chordCode } = props;
  const diagramDetails = generateFingerPosition(chordCode);

  return (
    <div className={diagramDetails.startFret}>
      <div className={diagramDetails.barre} />
      {diagramDetails.fingerPos}
    </div>
  );
}

ChordDiagram.defaultProps = {
  chordCode: '0,x,2,1,o,0,o',
};

ChordDiagram.propTypes = {
  chordCode: PropTypes.string,
};
