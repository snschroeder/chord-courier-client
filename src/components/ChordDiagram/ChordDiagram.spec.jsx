/* eslint-disable no-undef */
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import ChordDiagram from './ChordDiagram';

describe('ChordDiagram component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <ChordDiagram chordCode="0,x,2,1,o,0,o" />
      </BrowserRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders open position C maj chord when fed correct props', () => {
    const wrapper = shallow(<ChordDiagram chordCode="0,x,2,1,o,0,o" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
