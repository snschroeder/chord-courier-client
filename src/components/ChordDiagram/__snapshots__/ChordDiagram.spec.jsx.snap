// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChordDiagram component renders open position C maj chord when fed correct props 1`] = `
<div
  className="open-diagram"
>
  <div
    className=""
  />
  <MutedString
    key="1"
    stringNo={1}
  />
  <Finger
    fret={2}
    key="2"
    stringNo={2}
  />
  <Finger
    fret={1}
    key="3"
    stringNo={3}
  />
  <OpenString
    key="4"
    stringNo={4}
  />
  <Finger
    fret={0}
    key="5"
    stringNo={5}
  />
  <OpenString
    key="6"
    stringNo={6}
  />
</div>
`;
